velox/exec/Window.h

```diff
@@ -0,0 +1,228 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include "velox/exec/Operator.h"
+#include "velox/exec/RowContainer.h"
+#include "velox/exec/WindowFunction.h"
+#include "velox/exec/WindowPartition.h"
+
+namespace facebook::velox::exec {
+
+// This is a very simple in-Memory implementation of a Window Operator
+// to compute window functions.
+//
+// This operator uses a very naive algorithm that sorts all the input
+// data with a combination of the (partition_by keys + order_by keys)
+// to obtain a full ordering of the input. We can easily identify
+// partitions while traversing this sorted data in order.
+// It is also sorted in the order required for the WindowFunction
+// to process it.
+//
+// We will revise this algorithm in the future using a HashTable based
+// approach pending some profiling results.
+class Window : public Operator {
+ public:
+  Window(
+      int32_t operatorId,
+      DriverCtx* driverCtx,
+      const std::shared_ptr<const core::WindowNode>& windowNode);
+
+  void addInput(RowVectorPtr input) override;
+
+  RowVectorPtr getOutput() override;
+
+  bool needsInput() const override {
+    return !noMoreInput_;
+  }
+
+  void noMoreInput() override;
+
+  BlockingReason isBlocked(ContinueFuture* /* unused */) override {
+    return BlockingReason::kNotBlocked;
+  }
+
+  bool isFinished() override {
+    return finished_;
+  }
+
+ private:
+  // Structure for the window frame for each function.
+  struct WindowFrame {
+    const core::WindowNode::WindowType type;
+    const core::WindowNode::BoundType startType;
+    const core::WindowNode::BoundType endType;
+    const std::optional<column_index_t> startChannel;
+    const std::optional<column_index_t> endChannel;
+  };
+
+  // Helper function to create WindowFunction and frame objects
+  // for this operator.
+  void createWindowFunctions(
+      const std::shared_ptr<const core::WindowNode>& windowNode,
+      const RowTypePtr& inputType);
+
+  // Helper function to create the buffers for peer and frame
+  // row indices to send in window function apply invocations.
+  void createPeerAndFrameBuffers();
+
+  // Function to compute the partitionStartRows_ structure.
+  // partitionStartRows_ is vector of the starting rows index
+  // of each partition in the data. This is an auxiliary
+  // structure that helps simplify the window function computations.
+  void computePartitionStartRows();
+
+  // This function is invoked after receiving all the input data.
+  // The input data needs to be separated into partitions and
+  // ordered within it.
+  void sortPartitions();
+
+  // Helper function to call WindowFunction::resetPartition() for
+  // all WindowFunctions.
+  void callResetPartition(vector_size_t idx);
```

Comment:
What is 'idx'? Would it make sense to rename it to partition?

Reply:
Changed this to partitionNumber.